gcp/lib/src/gpc_region.dart

```diff
@@ -0,0 +1,104 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:io';
+
+import 'package:http/http.dart' as http;
+
+import 'bad_configuration_exception.dart';
+
+/// A convenience wrapper that first tries [regionFromEnvironment]
+/// then (if the value is `null`) tries [regionFromMetadataServer]
+///
+/// Like [regionFromMetadataServer], if no value is found, a
+/// [BadConfigurationException] is thrown.
+Future<String> computeRegion() async {
+  final localValue = regionFromEnvironment();
+  if (localValue != null) {
+    return localValue;
+  }
+  final result = await regionFromMetadataServer();
+
+  return result;
+}
+
+/// Returns the
+/// [Region](https://cloud.google.com/compute/docs/regions-zones#identifying_a_region_or_zone)
+/// for the current instance by checking the environment variables
+/// in [gcpRegionEnvironmentVariables].
+///
+/// The list is checked in order. This is useful for local development.
+///
+/// If no matching variable is found, `null` is returned.
+String? regionFromEnvironment() {
+  for (var envKey in gcpRegionEnvironmentVariables) {
+    final value = Platform.environment[envKey];
+    if (value != null) return value;
+  }
+
+  return null;
+}
+
+/// Returns a [Future] that completes with the
+/// [Region](https://cloud.google.com/compute/docs/regions-zones#identifying_a_region_or_zone)
+/// for the current instance by checking
+/// [instance metadata](https://cloud.google.com/compute/docs/metadata/default-metadata-values#vm_instance_metadata).
+///
+/// If the metadata server cannot be contacted, a [BadConfigurationException] is
+/// thrown.
+Future<String> regionFromMetadataServer() async {
+  const host = 'http://metadata.google.internal/';
+  final url = Uri.parse('$host/computeMetadata/v1/instance/region');
```

Comment:
We should DRY up this logic. Maybe add a `metadata_server.dart` file and share the logic with `gcp_project.dart`?

Reply:
@prefanatic thoughts on my thoughts?

Reply:
@kevmoo Sorry about the delay - I took a few stabs at drying up this implementation, but didn't like the direction I found myself in.

One attempt at buttoning this up included defining `MetadataServerClient` (name withstanding) which exposed these region and project endpoints. However, I was unhappy with aspects related to error handling; we disclose some nice resolution regarding which environment variables to provide in leu of the metadata server. It felt weird to me to move that into such a `metadata_server.dart`. 

In an attempt with leaving the error handling out, I ended up with an implementation that's slightly analogous to [http.read](https://pub.dev/documentation/http/latest/http/read.html), where `computeRegion` and `computeProject` would open up a client, try and return the metadata result, or catch into statement about environment variables needed. This slimmed up the original methods a little bit, but now introduced some arguably duplicative `withClient` semantics to retain backwards compatibility in those methods.

I _think_ this approach is worth taking? Let me know if I'm on the right track.

Reply:
What do you think of this? https://github.com/GoogleCloudPlatform/functions-framework-dart/pull/421

Reply:
Ah, looks much nicer. I wish I thought of using enhanced enums!

I left some small documentation nits over there for the generalization, in the review.